src/Npgsql/TaskExtensions.cs

```diff
@@ -83,5 +83,27 @@ internal static async Task WithCancellation(this Task task, CancellationToken ca
         internal static Task WithCancellationAndTimeout(this Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken)
             => task.WithCancellation(cancellationToken).WithTimeout(timeout);
 #endif
+
+        internal static TResult ExecuteWithCancellationAndTimeout<TResult>(Func<CancellationToken, TResult> func, NpgsqlTimeout timeout, CancellationToken cancellationToken)
+        {
+            CancellationTokenSource? combinedCts = null;
+            try
+            {
+                var finalCt = cancellationToken;
+
+                if (timeout.IsSet)
+                {
+                    combinedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                    combinedCts.CancelAfter((int)timeout.CheckAndGetTimeLeft().TotalMilliseconds);
+                    finalCt = combinedCts.Token;
+                }
+
+                return func(finalCt);
+            }
+            finally
+            {
+                combinedCts?.Dispose();
```

Comment:
This seems like a bug - the combinedCts is disposed when this method returns, which is immediately; so Dns.GetHostAddressesAsync hardly gets a chance to start running.

Reply:
Oh! Shame on me! What a noob mistake 🙄

Reply:
I do these all the time :) Everything is easy if all your methods are async, but the moment you start doing async with non-async methods, you have to always keep this in mind...